substra/sdk/backends/local/compute/worker.py

```diff
@@ -72,30 +125,103 @@ def _context(self, key):
+            input_models_volume = _mkdir(os.path.join(tuple_dir, "input_models"))
+            output_models_volume = _mkdir(os.path.join(tuple_dir, "output_models"))
+            if tuple_.in_head_model:
+                os.link(
+                    tuple_.in_head_model.storage_address,
+                    os.path.join(input_models_volume, tuple_.in_head_model.key)
+                )
+            if tuple_.in_trunk_model:
+                os.link(
+                    tuple_.in_trunk_model.storage_address,
+                    os.path.join(input_models_volume, tuple_.in_trunk_model.key)
+                )
 
             data_volume = self._get_data_volume(tuple_dir, tuple_)
 
             volumes = {
                 dataset.data_opener: _VOLUME_OPENER,
                 data_volume: _VOLUME_INPUT_DATASAMPLES,
+                input_models_volume: _VOLUME_INPUT_MODELS_RO,
+                output_models_volume: _VOLUME_OUTPUT_MODELS_RW
+            }
+
+            if tuple_.compute_plan_id:
+                local_volume = _mkdir(
+                    os.path.join(
+                        self._wdir, "compute_plans", "local", tuple_.compute_plan_id
+                    )
+                )
+                volumes[local_volume] = _VOLUME_LOCAL
+
+            # compute composite traintuple command
+            command = self._get_command_train_composite(
+                tuple_=tuple_,
+                models_volume=input_models_volume,
+                container_volume=_VOLUME_INPUT_MODELS_RO,
+            )
+
+            container_name = f"algo-{algo.key}"
+            logs = self._spawner.spawn(
+                container_name, str(algo.file), command, volumes=volumes
+            )
+
+            # save move output models
+            tuple_.out_head_model.out_model = self._save_output_model(
+                tuple_,
+                'output_head_model',
+                output_models_volume
+            )
+            tuple_.out_trunk_model.out_model = self._save_output_model(
+                tuple_,
+                'output_trunk_model',
+                output_models_volume
+            )
+
+            # set logs and status
+            tuple_.log = "\n".join(logs)
+            tuple_.status = models.Status.done
+
+            if compute_plan:
+                compute_plan.done_count += 1
+                if compute_plan.done_count == compute_plan.tuple_count:
+                    compute_plan.status = models.Status.done
+
+    def schedule_traintuple(self, tuple_):
+        """Schedules a ML task (blocking)."""
+        with self._context(tuple_.key) as tuple_dir:
+            tuple_.status = models.Status.doing
+
+            # fetch dependencies
+            if isinstance(tuple_, models.Traintuple):
+                algo = self._db.get(schemas.Type.Algo, tuple_.algo_key)
+            else:
+                algo = self._db.get(schemas.Type.AggregateAlgo, tuple_.algo_key)
```

Comment:
You could use the same strategy used for the `type_` class attribute and add an `algo_type` attribute so that you can write `algo = self._db.get(tuple_.__class__.algo_type, tuple_.algo_key)`

Reply:
Good, the only thing that is a bit weird is that here `tuple_` is of type `models.Traintuple` but in the backend, `tuple_` is of class `schemas.TraintupleSpec`, so I added the `algo_type` to the models, and the `compute_plan_attr_name` to the schemas.